In [2]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Plot parameters
sns.set()
# Seven hls color palette
current_palette_7 = sns.color_palette("hls", 7)
sns.set_palette(current_palette_7)

%pylab inline
pylab.rcParams['figure.figsize'] = (4, 4)
plt.rcParams['xtick.major.size'] = 0
plt.rcParams['ytick.major.size'] = 0
# rcParams.keys()

Populating the interactive namespace from numpy and matplotlib


In [4]:
# Avoid inaccurate floating values (for inverse matrices in dot product for instance)
# See https://stackoverflow.com/questions/24537791/numpy-matrix-inversion-rounding-errors
np.set_printoptions(suppress=True)

$$
\newcommand\norm[1]{\left\lVert#1\right\rVert} 
\DeclareMathOperator{\Tr}{Tr}
$$

# 2.10 The Trace Operator

The trace is the sum of all values in the diagonal of a square matrix.

$
\boldsymbol{A}=
\begin{bmatrix}
    2 & 9 & 8 \\\\
    4 & 7 & 1 \\\\
    8 & 2 & 5
\end{bmatrix}
$

$
\mathrm{Tr}(\boldsymbol{A}) = 2 + 7 + 5 = 14
$

The Numpy package provide the function `trace()` to calculate it:

In [5]:
A = np.array([[2, 9, 8], [4, 7, 1], [8, 2, 5]])
A_tr = np.trace(A)
A_tr

14

GoodFellow et al. explain that the trace can be used to specify the Frobenius norm of a matrix (see 2.5[]()). The Frobenius norm is the equivalent of the $L^2$ norm for matrices. It is defined by:

$
\norm{\boldsymbol{A}}_F=\sqrt{\sum_{i,j}A^2_{i,j}}
$

Take the square of all element and sum them. Take the square root of the result. This norm can also be calculated with:

$
\norm{\boldsymbol{A}}_F=\sqrt{\Tr({\boldsymbol{AA}^T})}
$

We can check this. The first way to compute the norm can be done with the simple command `np.linalg.norm()`:

In [11]:
np.linalg.norm(A)

17.549928774784245

The Frobenius norm of $\boldsymbol{A}$ is 17.549928774784245.

With the trace the result is identical:

In [13]:
np.sqrt(np.trace(A.dot(A.T)))

17.549928774784245

In [15]:
A.T

array([[2, 4, 8],
       [9, 7, 2],
       [8, 1, 5]])

Since the transposition of a matrix doesn't change the diagonal, the trace of the matrix is equal to the trace of its transpose:

$
\Tr(\boldsymbol{A})=\Tr(\boldsymbol{A}^T)
$

## Trace of a product

$
\Tr(\boldsymbol{ABC}) = \Tr(\boldsymbol{CAB}) = \Tr(\boldsymbol{BCA})
$


### Example 1.

Let's see an example of this property.

$
\boldsymbol{A}=
\begin{bmatrix}
    4 & 12 \\\\
    7 & 6
\end{bmatrix}
$

$
\boldsymbol{B}=
\begin{bmatrix}
    1 & -3 \\\\
    4 & 3
\end{bmatrix}
$

$
\boldsymbol{C}=
\begin{bmatrix}
    6 & 6 \\\\
    2 & 5
\end{bmatrix}
$



In [20]:
A = np.array([[4, 12], [7, 6]])
B = np.array([[1, -3], [4, 3]])
C = np.array([[6, 6], [2, 5]])

print np.trace(A.dot(B).dot(C))
print np.trace(C.dot(A).dot(B))
print np.trace(B.dot(C).dot(A))

531
531
531


$
\boldsymbol{ABC}=
\begin{bmatrix}
    360 & 432 \\\\
    180 & 171
\end{bmatrix}
$

$
\boldsymbol{CAB}=
\begin{bmatrix}
    498 & 126 \\\\
    259 & 33
\end{bmatrix}
$

$
\boldsymbol{BCA}=
\begin{bmatrix}
    -63 & -54 \\\\
    393 & 594
\end{bmatrix}
$

$
\Tr(\boldsymbol{ABC}) = \Tr(\boldsymbol{CAB}) = \Tr(\boldsymbol{BCA}) =  531
$

# References

[1] [Trace (linear algebra) - Wikipedia](https://en.wikipedia.org/wiki/Trace_(linear_algebra))

[2] [Numpy Trace operator](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trace.html)